File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> Result;
        set<vector<int>> s;
        for (int i = 0; i < nums.size() ; i++ ) {
            int m = i + 1;
            int j = nums.size() - 1;
            while (m < j) {
                int sum = nums[i] + nums[m] + nums[j];
                if (sum == 0){
                    s.insert({nums[i],nums[m],nums[j]});
                    m++;
                    j--;
                }
                else if (sum < 0) {
                    m++;
                }
                else{
                      j--;
                }
            }

        }
        for (auto allTriplets : s) 
            Result.push_back(allTriplets);
        return Result;
        

    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        length = len(nums)
        triplets = []
        for i in range(length -2):
            if i > 0 and  nums [i] == nums [i - 1]:
                continue

            left = i + 1 
            right = length - 1

            while  left < right:
                if nums [i] + nums [left] + nums [right] == 0:
                    triplets.append([nums[i], nums[left], nums [right]])
                    left = left  + 1 
                    right = right - 1  
                    while left < right and nums [left] == nums [left + 1 ]:
                        left = left + 1 
                    while left < right and nums [right] == nums [right - 1 ]:
                        right =  right - 1
                elif  nums [i] + nums [left] + nums [right] <= 0:
                    left = left + 1 
                else:
                    right = right - 1
        
        return triplets


